src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs

```diff
@@ -512,22 +513,53 @@ protected virtual void Dispose(bool disposing)
 
             if (disposing)
             {
-                foreach (var client in _clients)
+                if (!_disposedAsync)
                 {
-                    client.Dispose();
+                    DisposeAsync()
+                        .AsTask()
+                        .ConfigureAwait(false)
+                        .GetAwaiter()
+                        .GetResult();
                 }
+            }
 
-                foreach (var factory in _derivedFactories)
-                {
-                    factory.Dispose();
-                }
+            _disposed = true;
+        }
+
+        /// <inheritdoc />
+        public async ValueTask DisposeAsync()
```

Comment:
We want to make this virtual to allow people disposing their own resources in derived classes if needed.

Reply:
done